paddle/fluid/operators/prune_gate_by_capacity_op.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/operators/prune_gate_by_capacity_op.h"
+
+namespace paddle {
+namespace operators {
+
+class PruneGateByCapacityOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("GateIdx"), "Input", "GateIdx",
+                   "prun_gate_by_capacity");
+    OP_INOUT_CHECK(ctx->HasInput("ExpertCount"), "Input", "ExpertCount",
+                   "prun_gate_by_capacity");
+
+    OP_INOUT_CHECK(ctx->HasOutput("NewGateIdx"), "Output", "NewGateIdx",
+                   "prun_gate_by_capacity");
+    OP_INOUT_CHECK(ctx->HasOutput("ExpertCountOut"), "Output", "ExpertCountOut",
+                   "prun_gate_by_capacity");
+    auto gate_idx_dims = ctx->GetInputDim("GateIdx");
+    auto expert_count_dims = ctx->GetInputDim("ExpertCount");
+
+    int64_t n_expert = ctx->Attrs().Get<int64_t>("n_expert");
+    int64_t n_worker = ctx->Attrs().Get<int64_t>("n_worker");
+
+    int64_t gate_idx_num_ele = 1;
+    for (int64_t i = 0; i < gate_idx_dims.size(); i++) {
+      gate_idx_num_ele *= gate_idx_dims[i];
+    }
+    int64_t expert_count_num_ele = 1;
+    for (int64_t i = 0; i < gate_idx_dims.size(); i++) {
+      expert_count_num_ele *= expert_count_dims[i];
+    }
+
+    PADDLE_ENFORCE_EQ(
+        gate_idx_num_ele, n_expert * n_worker,
```

Comment:
How to get this conclusion？

Reply:
> How to get this conclusion？

In `fast_moe/cuda/tests/prune_gate.cu` line 12, 
```
int tot_expert = n_worker * n_expert;
long* gate_idx = new long[batch_size];
long* n_gate_idx = new long[batch_size];
``` 
I think the total elements of input gate_id are the product of n_worker and n_expert.